com.kms.cura.entity/src/com/kms/cura/entity/MessageThreadEntity.java

```diff
@@ -0,0 +1,110 @@
+        ArrayList<MessageThreadEntity> messageThreadEntities = new ArrayList<>();
+        if (entity instanceof PatientUserEntity) {
+            for (int i = 0; i < messageEntities.size(); ++i) {
+                MessageEntity messageEntity = messageEntities.get(i);
+                if (messageEntity.isSenderDoctor()) {
+                    int exist = checkExist(messageEntity.getSender().getName(), entity, messageThreadEntities);
+                    if (exist == -1) {
+                        MessageThreadEntity messageThreadEntity = new MessageThreadEntity();
+                        messageThreadEntity.addMessage(messageEntity);
+                        messageThreadEntities.add(messageThreadEntity);
+                    } else {
+                        messageThreadEntities.get(exist).addMessage(messageEntity);
+                    }
+                } else {
+                    int exist = checkExist(messageEntity.getReceiver().getName(), entity, messageThreadEntities);
+                    if (exist == -1) {
+                        MessageThreadEntity messageThreadEntity = new MessageThreadEntity();
+                        messageThreadEntity.addMessage(messageEntity);
+                        messageThreadEntities.add(messageThreadEntity);
+                    } else {
+                        messageThreadEntities.get(exist).addMessage(messageEntity);
+                    }
+                }
+            }
+        }
+        if (entity instanceof DoctorUserEntity) {
+            for (int i = 0; i < messageEntities.size(); ++i) {
+                MessageEntity messageEntity = messageEntities.get(i);
+                if (messageEntity.isSenderDoctor()) {
+                    int exist = checkExist(messageEntity.getReceiver().getName(), entity, messageThreadEntities);
+                    if (exist == -1) {
+                        MessageThreadEntity messageThreadEntity = new MessageThreadEntity();
+                        messageThreadEntity.addMessage(messageEntity);
+                        messageThreadEntities.add(messageThreadEntity);
+                    } else {
+                        messageThreadEntities.get(exist).addMessage(messageEntity);
+                    }
+                } else {
+                    int exist = checkExist(messageEntity.getSender().getName(), entity, messageThreadEntities);
+                    if (exist == -1) {
+                        MessageThreadEntity messageThreadEntity = new MessageThreadEntity();
+                        messageThreadEntity.addMessage(messageEntity);
+                        messageThreadEntities.add(messageThreadEntity);
+                    } else {
+                        messageThreadEntities.get(exist).addMessage(messageEntity);
+                    }
+                }
+            }
+        }
+        return messageThreadEntities;
```

Comment:
This method is too complicated and have a lot of duplicated code. Should try to re-factor it.